File: backend/api/quick_actions.py
from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, status

from app.ai.errors import AIConfigurationError, AIError
from backend.api.dependencies import get_quick_action_service
from backend.models.quick_actions import (
    QuickActionRequest,
    QuickActionResponse,
    QuickActionStatusResponse,
)
from backend.services.quick_action_service import QuickActionService

router = APIRouter(prefix="/api/quick-actions", tags=["quick-actions"])
QuickActionServiceDependency = Annotated[
    QuickActionService,
    Depends(get_quick_action_service),
]


@router.get("/status", response_model=QuickActionStatusResponse)
def quick_action_status(service: QuickActionServiceDependency) -> QuickActionStatusResponse:
    available, provider, model, detail = service.status()
    return QuickActionStatusResponse(
        available=available,
        provider=provider,
        model=model,
        detail=detail,
    )


@router.post("/run", response_model=QuickActionResponse)
def run_quick_action(
    payload: QuickActionRequest,
    service: QuickActionServiceDependency,
) -> QuickActionResponse:
    try:
        result = service.run(**payload.model_dump())
    except AIConfigurationError as exc:
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail=str(exc),
        ) from exc
    except AIError as exc:
        raise HTTPException(
            status_code=status.HTTP_502_BAD_GATEWAY,
            detail=str(exc),
        ) from exc

    return QuickActionResponse(
        action=result.action,
        output_text=result.output_text,
        provider=result.provider,
        model=result.model,
        request_id=result.request_id,
    )

File: apps/desktop/src/api/quick-actions.ts
import { apiGet, apiPost } from "./client"
import type {
  QuickActionRequest,
  QuickActionResponse,
  QuickActionStatusResponse,
  ResearchNoteListResponse,
  ResearchNoteSaveRequest,
  ResearchNoteSaveResponse,
} from "./types"

export function getQuickActionStatus(): Promise<QuickActionStatusResponse> {
  return apiGet<QuickActionStatusResponse>("/api/quick-actions/status")
}

export function runQuickAction(payload: QuickActionRequest): Promise<QuickActionResponse> {
  return apiPost<QuickActionResponse, QuickActionRequest>("/api/quick-actions/run", payload)
}

export function saveResearchNote(payload: ResearchNoteSaveRequest): Promise<ResearchNoteSaveResponse> {
  return apiPost<ResearchNoteSaveResponse, ResearchNoteSaveRequest>("/api/research/notes", payload)
}

export function listResearchNotes(limit = 5): Promise<ResearchNoteListResponse> {
  return apiGet<ResearchNoteListResponse>(`/api/research/notes?limit=${limit}`)
}
